core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java

```diff
@@ -937,13 +941,28 @@ public void onFailure(Throwable throwable)
             }, directExecutor());
         }
         else {
+            committed.set(true);
             transitionToFinished();
         }
         return true;
     }
 
+    public void resultsConsumed()
+    {
+        consumed.set(true);
+        transitionToFinished();
```

Comment:
transtionToFailed now has conditional logic inside and may or may not do actual transition.
Maybe renamt to `transtionToFinishedIfReady`?